javagen/src/main/java/com/azure/autorest/template/ClientMethodTemplate.java

```diff
@@ -48,111 +52,189 @@ public static ClientMethodTemplate getInstance() {
         return INSTANCE;
     }
 
-    protected static void addValidations(JavaBlock function, List<String> expressionsToCheck, Map<String, String> validateExpressions, JavaSettings settings) {
-        if (settings.shouldClientSideValidations()) {
-            for (String expressionToCheck : expressionsToCheck) {
-                JavaIfBlock nullCheck = function.ifBlock(expressionToCheck + " == null", ifBlock ->
-                        ifBlock.line("return Mono.error(new IllegalArgumentException(\"Parameter %s is required and "
-                            + "cannot be null.\"));", expressionToCheck));
-                if (validateExpressions.containsKey(expressionToCheck)) {
-                    nullCheck.elseBlock(elseBlock ->
-                            elseBlock.line(validateExpressions.get(expressionToCheck) + ";"));
-                }
+    /**
+     * Adds validations to the client method.
+     *
+     * @param function The client method code block.
+     * @param expressionsToCheck Expressions to validate as non-null.
+     * @param validateExpressions Expressions to validate with a custom validation (key is the expression, value is the
+     * validation).
+     * @param settings AutoRest generation settings, used to determine if validations should be added.
+     */
+    protected static void addValidations(JavaBlock function, List<String> expressionsToCheck,
+        Map<String, String> validateExpressions, JavaSettings settings) {
+        if (!settings.shouldClientSideValidations()) {
+            return;
+        }
+
+        // Iteration of validateExpressions uses expressionsToCheck effectively as a set lookup, may as well turn the
+        // expressionsToCheck to a set.
+        Set<String> expressionsToCheckSet = new LinkedHashSet<>(expressionsToCheck);
+
+        for (String expressionToCheck : expressionsToCheckSet) {
+            // TODO (alzimmer): Determine if the assumption being made here are always true.
+            // 1. Assumes that the expression is nullable.
+            // 2. Assumes that the client method returns a reactive response.
+            // 3. Assumes that the reactive response is a Mono.
+            JavaIfBlock nullCheck = function.ifBlock(expressionToCheck + " == null", ifBlock ->
+                ifBlock.line("return Mono.error(new NullPointerException(\"Parameter %s is required and "
```

Comment:
The change to `NullPointerException` would be a big behavior change. And this could be kind of inconsistency depends on when the code regened...

Reply:
And if you really need to change this, remember to update javadoc as well (I am not sure whether currently we had any validation other than null check).
```
     * @throws IllegalArgumentException thrown if parameters fail the validation.
```

Reply:
I think it is best to revert this back to `IllegalArgumentException` for now given the effect this change could have, we'll want to have a more targeted discussion for this.